core/src/com/unciv/logic/civilization/managers/TechManager.kt

```diff
@@ -369,12 +369,17 @@ class TechManager : IsPartOfGameInfoSerialization {
                 .toMutableList()
         }
 
-        // As long as TurnManager does cities after tech, we don't need to clean up
-        // inProgressConstructions - CityConstructions.validateInProgressConstructions does it.
-
         // Add notifications for obsolete units/constructions
         for ((unit, cities) in unitUpgrades) {
             if (cities.isEmpty()) continue
+
+            //The validation check happens again while processing start and end of turn,
+            //but for mid-turn free tech picks like Oxford University, it should happen immediately
+            //so the hammers from the obsolete unit are guaranteed to go to the upgraded unit
+            //and players don't think they lost all their production mid turn
+            for(c in cities)
+                c.cityConstructions.validateInProgressConstructions()
```

Comment:
```suggestion
            for (city in cities)
                city.cityConstructions.validateInProgressConstructions()
```

Reply:
This should go after line 369, since that's where we deal with rebuilding the construction queue it's also where we should validate it

Reply:
I thought about adding it to that loop, but that loop is over every single city. I mean the wasted checks aren't really going to affect performance much because the number of cities and depth of inProgressConstructions is probably never going to be too large, but it felt inefficient?

I thought about putting it in the transform function, but it is possible a single city could have two obsolete units from the same tech so it would require replacing validateInProgressConstructions with a new function that handles one unit replacement at a time (That replacement seems more logical if we can assume that tech is the only way to obsolete a unit, then we could stop calling validate function every start and end of turn and just only handle it when tech upgrades occur).

I am up for any of the options, you choose and I will change it:

1. leave where it is (I will fix the loop variable name)
2. move call to line 369 and run it on every city
3. rewrite validateInProgressConstructions to only happen on tech upgrades and not on every turn (probably replace it with a new function something like 'handleObsoleteConstruction(oldConstruction, newConstruction)' that does the queue replacement and inProgressConstruction transfer all at once)

Reply:
I see, this is fine